text/0000-nested-publish.md

```diff
@@ -0,0 +1,183 @@
+- Feature Name: `nested_publish`
+- Start Date: 2023-06-30
+- RFC PR: [rust-lang/rfcs#3452](https://github.com/rust-lang/rfcs/pull/3452)
+- Rust Issue: ...
+
+# Summary
+[summary]: #summary
+
+Allow Cargo packages to be bundled within other Cargo packages when they are published (not just in unpublished workspaces).
+
+# Motivation
+[motivation]: #motivation
+
+There are a number of reasons why a Rust developer currently may feel the need to create multiple library crates, and therefore multiple Cargo packages (since one package contains at most one library crate). These multiple libraries could be:
+
+* A trait declaration and a corresponding derive macro (which must be defined in a separate proc-macro library).
+* A library that uses a build script that uses another library or binary (e.g. for precomputation or bindings generation).
+* A logically singular library broken into multiple parts to speed up compilation.
+
+Currently, developers must publish these packages separately. This has several disadvantages (see the [Rationale](#rationale-and-alternatives) section for further details):
+
+* Clutters the public view of the registry with packages not intended to be usable on their own, and which may even become obsolete as internal architecture changes.
+* Requires multiple `cargo publish` operations (this could be fixed with bulk publication) and writing public metadata for each package.
+* Can result in semver violations and thus compilation failures, due to the developer not thinking about semver compatibility within the group.
+
+This RFC will allow developers to avoid all of these inconveniences and hazards by publishing a single package.
+
+# Guide-level explanation
+[guide-level-explanation]: #guide-level-explanation
+
+By default (and always, prior to this RFC's implementation):
+
+* If your package contains any sub-packages, Cargo [excludes](https://doc.rust-lang.org/cargo/reference/manifest.html#the-exclude-and-include-fields) them from the `.crate` archive file produced by `cargo package` and `cargo publish`.
+* If your package contains any non-`dev` dependencies which do not give a `version = "..."`, it cannot be published to `crates.io`.
+
+(By “**sub-package**” we mean a package (directory with `Cargo.toml`) which is a subdirectory of another package. We shall call the outermost such package, the package being published, the “**parent package**”.)
+
+You can change this default by placing in the manifest (`Cargo.toml`) of a sub-package:
+
+```toml
+[package]
+publish = "nested"
+```
+
+If this is done, Cargo's behavior changes as follows:
+
+* If you publish the parent package, the sub-package is included in the `.crate` file (unless overridden by explicit `exclude`/`include`) and will be available to the parent package whenever the parent package is downloaded and compiled.
+* The parent package may have a `path =` dependency upon the sub-package. (This dependency may not have a `version =` specified.)
```

Comment:
I would instead word this as

- If you publish a parent package with a path dependency to this sub-package, it will be included in the `.crate` file
- The dependency on the sub-package may not have a "registry" dependency (e.g. `version` field)

Reply:
> If you publish a parent package with a path dependency to this sub-package, it will be included in the `.crate` file

That's not just a different explanation, that's a different behavior. I intentionally kept “what files are published” decoupled from “is there an edge in the dependency graph”; this means that the file inclusion rules are kept simple and local. (Also, if the future possibility of public sub-packages comes to pass, then there would not be such a dependency edge, so there would need to be another rule to include those.)

I don't think it's a good idea to DWIM file inclusion too much because it can already go wrong. I've personally accidentally published a compiled binary inside a package because of some of its choices (not related to subpackages).

That said, I see the logic behind including packages only if they will be actually used. I'm just not yet convinced it's wise.

Reply:
- We should be somewhat mindful of bulk.  This could especially be bad in workspaces with a root package as that root package will pick up *every(* sub-package in the workspace.
- Maybe its just me but I feel like the logic for inclusion (implementation and in education) is much nicer when we are limiting it to direct path dependencies rather than
  - implementation:  "scan every workspace member and see if its a child"
  - education: dealing with "distant" effects (magically picking things up despite there being no explicit relationship)

Reply:
The root package scenario is convincing to me that we need it to be dependency-aware. I will think about how to revise the RFC; this has a lot of potential consequences.

Reply:
I have now updated the RFC text to specify that nested publishing requires `dependencies.*.publish = "nested"` in addition to the previous requirements before it will activate.